pkgs/development/beam-modules/build-mix.nix

```diff
@@ -1,100 +1,70 @@
+  };
 
-  pkg = self: stdenv.mkDerivation ( attrs // {
-    name = "${name}-${version}";
-    inherit version;
+  pkg = self:
+    stdenv.mkDerivation (attrs // {
+      name = "${name}-${version}";
+      inherit version;
 
-    dontStrip = true;
+      dontStrip = true;
 
-    inherit src;
+      inherit src;
 
-    setupHook = if setupHook == null
-    then writeText "setupHook.sh" ''
-       addToSearchPath ERL_LIBS "$1/lib/erlang/lib"
-    ''
-    else setupHook;
+      nativeBuildInputs = nativeBuildInputs ++ [ erlang hex elixir ];
 
-    inherit buildInputs;
-    propagatedBuildInputs = [ hex elixir ] ++ beamDeps;
+      MIX_ENV = mixEnv;
+      MIX_DEBUG = if enableDebugInfo then 1 else 0;
+      HEX_OFFLINE = 1;
+      DEBUG = if enableDebugInfo then 1 else 0; # for Rebar3 compilation
 
-    configurePhase = if configurePhase == null
-    then ''
-      runHook preConfigure
-      ${erlang}/bin/escript ${bootstrapper}
-      runHook postConfigure
-    ''
-    else configurePhase ;
+      postUnpack = ''
+        export HEX_HOME="$TMPDIR/hex"
+        export MIX_HOME="$TMPDIR/mix"
+        export MIX_DEPS_PATH="$TMPDIR/deps"
 
+        # Rebar
+        # the api with `mix local.rebar rebar path` makes a copy of the binary
+        export MIX_REBAR "${rebar}/bin/rebar"
+        export MIX_REBAR3 "${rebar3}/bin/rebar3"
+        export REBAR_GLOBAL_CONFIG_DIR="$TMPDIR/rebar3"
+        export REBAR_CACHE_DIR="$TMPDIR/rebar3.cache"
 
-    buildPhase = if buildPhase == null
-    then ''
-        runHook preBuild
+        cp --no-preserve=mode -R "${mixDeps}" "$MIX_DEPS_PATH"
+      '' + (attrs.postUnpack or "");
 
-        export HEX_OFFLINE=1
-        export HEX_HOME=`pwd`
-        export MIX_ENV=prod
-        export MIX_NO_DEPS=1
+      configurePhase = attrs.configurePhase or ''
+        runHook preConfigure
 
-        mix compile ${debugInfoFlag} --no-deps-check
+        mix deps.loadpaths
+        runHook postConfigure
+      '';
 
+      buildPhase = attrs.buildPhase or ''
+        runHook preBuild
+        mix do compile --no-deps-check, release --path "$out"
         runHook postBuild
-    ''
-    else buildPhase;
-
-    installPhase = if installPhase == null
-    then ''
-        runHook preInstall
-
-        MIXENV=prod
-
-        if [ -d "_build/shared" ]; then
-          MIXENV=shared
-        fi
-
-        mkdir -p "$out/lib/erlang/lib/${name}-${version}"
-        for reldir in src ebin priv include; do
-          fd="_build/$MIXENV/lib/${name}/$reldir"
-          [ -d "$fd" ] || continue
-          cp -Hrt "$out/lib/erlang/lib/${name}-${version}" "$fd"
-          success=1
-        done
-
-        runHook postInstall
-    ''
-    else installPhase;
-
-    passthru = {
-      packageName = name;
-      env = shell self;
-      inherit beamDeps;
-    };
-});
-in fix pkg
+      '';
+
+      dontInstall = true;
+
+      passthru = {
+        packageName = name;
+        env = shell self;
+      };
+    });
+in pkg lib.fix
```

Comment:
I just copied that from before. I have to say I'm not sure. I tested both ways and it doesn't seem to make a difference. We'll have to ask somebody more knowledgeable here.

Reply:
From my understanding, this allows the use of `self`, which is used only in `passthru.env`. I can confirm that trying to build `mobilizon.env` with `pkg lib.fix` produces an evaluation error, while with `lib.fix pkg` it builds but fails anaway with `variable $src or $srcs should point to the source`. In my opinion, we should drop the `passthru` and `lib.fix pkg` for now and open another PR if the need arises.

Reply:
Isn't the passthru stuff used for potentially passing the shell ? I'm not sure, just wondering if by removing those, we are loosing the ability to define ourselves the shell so that the user doesn't need do

Reply:
I am no sure, as far as I know, nix-shell for a package has always worked for me, whether the package defines an `env`. But I am not proposing to remove them indefinitely, though. I think defining the shell there is nice, because we can easily extend it using all the attributes of the package.